alternative onoff cancellation approach

pabigot opened the pull request:
This derives from the commits #23229 excluding the notifier, but uses a technique similar to that of the notifier to allow cancellation of in-progress operations without a failure case.  Cancelling a request or a reset always completes synchronously, even if this means errors in the transition are not reported.  Canceling a release converts it into a request that will force a transition to on as soon as the transition to stop completes.

nordic-krch committed:
lib: os: onoff: Move transition functions out of service struct

Extracted transition functions from onoff structure to external one
which allows to keep them in flash.
Added ONOFF_SERVICE_DEFINE macro for static initialization.

Signed-off-by: Krzysztof Chruscinski <krzysztof.chruscinski@nordicsemi.no>

pabigot committed:
lib: async_notify: add generic asynchronous notification infrastructure

k_poll() for a signal is often desired for notification of completion
of asynchronous operations, but there are APIs where it may be
necessary to invoke "asynchronous" operations from contexts where
sleep is disallowed, or before the kernel has been initialized.
Extract the general notification solution from the on-off service into
a utility that can be used for other APIs.

Also move documentation out to a resource management section.

Signed-off-by: Peter Bigot <peter.bigot@nordicsemi.no>

pabigot committed:
lib: os: onoff: generalize and shorten API

The original API was misnamed, as the intent was to provide a manager
that decoupled state management from the service that needed to be
turned on or off.  Update all the names, shortening them where
appropriate removing unncessary internal components like _service.

Also remove some API that misled developers into believing that onoff
managers are normally expected to be exposed directly to consumers.
While this is a use case, in most situations there are service or
client-specific actions that need to be coupled to transition events.

Signed-off-by: Peter Bigot <peter.bigot@nordicsemi.no>

pabigot committed:
lib: os: onoff: rework to support cancelling in-progress operations

An original design goal of the onoff infrastructure was to ensure that
any errors discovered during transitions would be communicated to at
least one client.  To meet this goal cancellation of the last client
associated with an in-progress transition must be rejected.

This ends up significantly complicating the logic associated with
operation cancellation, and impacted clients that were forced to
maintain state to issue a release or request when the in-progress
transition completed.

Rework transition management so that both request and release
transitions can be accomodated at the same time.  Leave a cancelled
request notification uncompleted, as the client can tell by the
cancellation return value that it has received control of the client
object.

Signed-off-by: Peter Bigot <peter.bigot@nordicsemi.no>

nordic-krch commented:
> Cancelling a request or a reset always completes synchronously, even if this means errors in the transition are not reported

So far this was a must have to get notification about an error and now it's not? Maybe there should be an option to have one monitor client which is notified about error `onoff_monitor(srv,cli)`.
In general client will learn that service is in error state on next request so that seems ok but something is needed for cases when it is critical to react to error on time.

nordic-krch commented on lib/os/onoff.c:
```diff
@@ -390,128 +408,151 @@ int onoff_release(struct onoff_service *srv,
 
 	if (rv == 0) {
-		notify_one(srv, cli, -ECANCELED);
+		notify_one(mgr, cli, -ECANCELED);
```
this notification is now somehow artificial and redundant. If cancellation returning 0 tells user that request/release is cancelled. No need to handle that again in the callback. Also that reduces a chance of error code conflict (service returning -ECANCELED error).

pabigot replied:
That was always intended so there would be exactly one place in the code to process the completion that confirms that the client structure has been released.  Not completing might make a client more complex.  I agree about deconflicting with the service error space, though, so this will probably change.

pabigot commented:
> So far this was a must have to get notification about an error and now it's not?

It was a design goal when cancellation was an highly unusual operation.  Since cancellation appears to be common at least with your use style, its cost has increased.  There are several ways forward.

In general I'm not in favor of having a client that was configured to receive notification of a request completion suddenly have to cope with receiving notification of a release completion when the request is cancelled.  I can probably live with that, though, with a little documentation update.

> Maybe there should be an option to have one monitor client which is notified about error `onoff_monitor(srv,cli)`.

Yes, something like that, though I was thinking of a registry of callbacks for anything interested in receiving notifications of state changes, probably by moving that feature out of `onoff_notifier`.  Again, more thought required.

I'll make a second pass on this then.

pabigot committed:
lib: os: onoff: rework to make state flags visible to clients

Upcoming changes will inform the clients of the state of the system at
points where it changes.  Make the state-related bits visible to
enable the clients to determine the current state.

Signed-off-by: Peter Bigot <peter.bigot@nordicsemi.no>

pabigot committed:
lib: os: onoff: include state at time of completion in notify callback

Upcoming revision of notifications for cancelled operations includes
the possibility that a client originally submitted for a request
operation will instead receive a notification for a completed release
operation.  Although this doesn't solve the problem for spinwait or
signal notifications, add the state-at-completion as a parameter to
the callback notification signature.

Since we're changing things, move the user data parameter to the last
argument, which is its position in most other callback APIs.

Signed-off-by: Peter Bigot <peter.bigot@nordicsemi.no>

pabigot committed:
lib: os: onoff: add ability to monitor service state changes

It's possible to have a transition complete with no clients available
to receive the result, which may be a notification of error.
Alternatively a notification that was originally associated with a
release may be converted to a notification associated with a request
due the cancellation of the release.  There is no way to detect this
from spinwait or signal notifications.

Also a module may want to centralize tracking of current state and
error handling for an onoff service that gets used from multiple
contexts.

In support of this provide API to monitor state changes within an
onoff service.  Include notifications when the service transitions out
of stable (ON or OFF) states as well as detection of errors.

Signed-off-by: Peter Bigot <peter.bigot@nordicsemi.no>

pabigot commented:
@nordic-krch Next pass at this.

Cancellation no longer includes completion with `-ECANCELLED`.  The notify callback now includes the state at time of completion, so that clients originally submitted for a release but due to cancellation used for a request get told the state that was delivered.  There's also a monitor facility to track changes to the service state, which is the only way to be notified that transition has started, and the only way to try to correlate the result state for cancelled releases for spinwait and signal.  See the commit messages.

This is now nearly equal to the capabilities of #23229.  It's not equivalent because there's noting comparable to onoff_notifier_request/release to manage the client-specific state.  I'm not entirely sure that cancellation is the right way to solve that, but will look at it tomorrow.

nordic-krch commented on include/sys/onoff.h:
```diff
@@ -568,44 +565,119 @@ int onoff_service_reset(struct onoff_service *srv,
+ * but must not register or unregister any other monitors.
+ *
+ * @param mgr the manager for which a transition has completed.
+ *
```
param monitor missing

nordic-krch commented on include/sys/onoff.h:
```diff
@@ -568,44 +565,119 @@ int onoff_service_reset(struct onoff_service *srv,
+ */
+typedef void (*onoff_monitor_callback)(struct onoff_manager *mgr,
+				       struct onoff_monitor *mon,
+				       u32_t state,
```
does it report also transient states (to off, to on)? if not maybe it could be just bool (true means on)? Then ONOFF_STATE_MASK could be private in source file.

pabigot replied:
Yes, it reports initiation of a transition.  This can be important because at that point the service can't be used anymore; somebody might watch that and choose to not initiate a request if it's known the service is shutting down.

State mask also covers the error bit, which is public.

nordic-krch reviewed (commented):
just briefly looked at the api

pabigot commented:
Closing: this approach is now integrated into the development branch available for pre-PR review as #23229.
